### PR TITLE
add compiler tools info

### PR DESCRIPTION
**Pull request by electronwill**
add compiler tools info

### CODE REVIEW

**Comment by electronwill** [comment]
@msarahan this is a first draft for moving the information in https://docs.google.com/document/d/103QY204eD82fJzpT7eXy_rV2bKwe6Hx5uAnkMg_xwkI/ into the docs. I do have a few questions:

1. "Compare with earlier available compiler packages (gcc on mac and linux), the new packages are split up into separate compilers:"

The earlier available compiler packages were not split?

2. "Again, this new _sysconfigdata customization scheme is added in recent versions of the Python package: you should be aware that if you’re not using conda-build to play with the new compilers, you may need to update to a newer Python package so that you have the correct _sysconfigdata files."

If they're not using conda-build to play with the new compilers, or if they are using conda-build to play with the new compilers?

3. "you may also need to use the new “host” section in the requirements section:"

What does setting host to zlib mean exactly? Is it telling conda-build to use the zlib in the conda environment and not the system zlib, so conda-build uses zlib for the target platform and not zlib for the build platform?

4. Are the line breaks correct in the long blocks of code? Do those comments end with "Wl," ?

5. "Note that we make the “meat” package clang agree in version with our output version (implicitly the same as the top-level recipe)."

Does that mean the same thing as this?

"Note that we make the “meat” package clang agree in version with our output version. This is implicitly the same as the top-level recipe."

File: docs/source/user-guide/tasks/build-packages/compiler-tools.rst [hunk 1]
```diff
@@ -0,0 +1,328 @@
...
+macOS:
+
+* clang_osx-64
+* clangxx_osx-64
+* gfortran_osx-64
+
+A compiler's "build platform" is the platform where the compiler runs and
+builds the code.
+
+A compiler's "target platform" is the platform where the built code will run.
```
Comment (mingwandroid) [inline docs/source/user-guide/tasks/build-packages/compiler-tools.rst]: Can we change "target platform" to "host platform"? We use "host" for this everywhere else and I think using "target" for some bits just muddies the water.

Comment (electronwill) [reply docs/source/user-guide/tasks/build-packages/compiler-tools.rst]: Oh! Yes, that's much better. I realized that we needed some simple terms for those concepts but didn't realize that host was being used already. (I got the build/target terminology from https://en.wikipedia.org/wiki/Cross_compiler . "The fundamental use of a cross compiler is to separate the build environment from target environment.") I'll edit now and replace build/target with build/host in all cases.

**Comment by electronwill** [comment]
@mingwandroid (and optionally @msarahan ): The Google doc includes the sentence "The latest Python package builds now have a default _sysconfigdata file that keeps the default compilers as the ones (gcc, g++, etc.) provided by the host system." Should this be "build system"?

**Comment by electronwill** [comment]
Changed "target" to "host". "target_platform" is still used 8 places in the code, and I'm not sure if we should change that too.

File: docs/source/user-guide/tasks/build-packages/compiler-tools.rst (hunk 1 above)
Comment (mingwandroid) [reply docs/source/user-guide/tasks/build-packages/compiler-tools.rst]: To help define "host" in the readers' heads I'd be tempted to go for:

> A compiler's "host platform" is the platform where the built code will finally be hosted/run.

Comment (electronwill) [reply docs/source/user-guide/tasks/build-packages/compiler-tools.rst]: I like that. Went ahead and made the change.

**Comment by msarahan** [comment]
> The earlier available compiler packages were not split?

No, the gcc 4.8.5 packages include gcc, g++ and gfortran all in one.

> If they're not using conda-build to play with the new compilers, or if they are using conda-build to play with the new compilers?

Conda-build will automatically try to use the latest python version available in the currently configured channels.  Hopefully that gets the latest from defaults.  If they're not using conda-build, conda will not be auto-updating python, and they may have an older version.

> What does setting host to zlib mean exactly? Is it telling conda-build to use the zlib in the conda environment and not the system zlib, so conda-build uses zlib for the target platform and not zlib for the build platform?

yes, you have the right idea.

> Are the line breaks correct in the long blocks of code? Do those comments end with "Wl," ?

They seem OK to me by looking at the rendered version.  If there's bugs, it's ok, we can fix them in a less complicated follow-up.

> Does that mean the same thing as this?

Yes, that's a good rephrasing that improves readability.

**Comment by msarahan** [comment]
target_platform is unfortunately something that conda-build treats specially.  It should not be changed without a corresponding change in conda-build, and I think that ship has sailed.

**Comment by electronwill** [comment]
Thank you both very much!

That answers everything, except for double checking that this bit is correct:

"The latest Python package builds now have a default _sysconfigdata file that keeps the default compilers as the ones (gcc, g++, etc.) provided by the host system.  This hopefully keeps legacy recipes working."

Is it correct for that to be "provided by the host system" and not "provided by the build system"?

**Comment by mingwandroid** [comment]
I'd just use system here I think.

**Comment by electronwill** [comment]
"provided by the system" sounds perfect. Thanks.

**Comment by electronwill** [comment]
Unless anyone disagrees I'll merge this when CI passes.

**Pull request merged**